core/services/registrysyncer/syncer.go

```diff
@@ -0,0 +1,253 @@
+
+	kcr "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/capabilities_registry"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+	evmrelaytypes "github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+type HashedCapabilityID [32]byte
+type DonID uint32
+
+type State struct {
+	IDsToDONs         map[DonID]kcr.CapabilitiesRegistryDONInfo
+	IDsToNodes        map[p2ptypes.PeerID]kcr.CapabilitiesRegistryNodeInfo
+	IDsToCapabilities map[HashedCapabilityID]kcr.CapabilitiesRegistryCapabilityInfo
+}
+
+type Launcher interface {
+	Launch(ctx context.Context, state State) error
+}
+
+type Syncer interface {
+	services.Service
+	AddLauncher(h ...Launcher)
+}
+
+type registrySyncer struct {
+	stopCh    services.StopChan
+	launchers []Launcher
+	reader    types.ContractReader
+
+	wg   sync.WaitGroup
+	lggr logger.Logger
+	mu   sync.RWMutex
+}
+
+var _ services.Service = &registrySyncer{}
+
+var (
+	defaultTickInterval = 12 * time.Second
+)
+
+// New instantiates a new RegistrySyncer
+func New(
+	lggr logger.Logger,
+	relayer contractReaderFactory,
+	registryAddress string,
+) (*registrySyncer, error) {
+	stopCh := make(services.StopChan)
+	ctx, _ := stopCh.NewCtx()
+	reader, err := newReader(ctx, lggr, relayer, registryAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	return newSyncer(
+		stopCh,
+		lggr.Named("RegistrySyncer"),
+		reader,
+	), nil
+}
+
+type contractReaderFactory interface {
+	NewContractReader(context.Context, []byte) (types.ContractReader, error)
+}
+
+func newReader(ctx context.Context, lggr logger.Logger, relayer contractReaderFactory, remoteRegistryAddress string) (types.ContractReader, error) {
+	contractReaderConfig := evmrelaytypes.ChainReaderConfig{
+		Contracts: map[string]evmrelaytypes.ChainContractReader{
+			"CapabilitiesRegistry": {
+				ContractABI: kcr.CapabilitiesRegistryABI,
+				Configs: map[string]*evmrelaytypes.ChainReaderDefinition{
+					"getDONs": {
+						ChainSpecificName: "getDONs",
+					},
+					"getCapabilities": {
+						ChainSpecificName: "getCapabilities",
+					},
+					"getNodes": {
+						ChainSpecificName: "getNodes",
+					},
+				},
+			},
+		},
+	}
+
+	contractReaderConfigEncoded, err := json.Marshal(contractReaderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cr, err := relayer.NewContractReader(ctx, contractReaderConfigEncoded)
+	if err != nil {
+		return nil, err
+	}
+
+	err = cr.Bind(ctx, []types.BoundContract{
+		{
+			Address: remoteRegistryAddress,
+			Name:    "CapabilitiesRegistry",
+		},
+	})
+
+	return cr, err
+}
+
+func newSyncer(
+	stopCh services.StopChan,
+	lggr logger.Logger,
+	reader types.ContractReader,
+) *registrySyncer {
+	return &registrySyncer{
+		stopCh: stopCh,
+		lggr:   lggr,
+		reader: reader,
+	}
+}
+
+func (s *registrySyncer) Start(ctx context.Context) error {
+	s.mu.RLock()
```

Comment:
the read lock seems to be wrong. what is it supposed to be protecting?

Reply:
Ah correct :) We used to range of launchers here, but don't anymore. I'll remove it.